### PR TITLE
support abstract classes for resteasy #12927

### LINKED ISSUE #12927: RESTEasy tries to instantiate abstract resource base class
**Describe the bug**
If your REST resource extends an abstract base class that contains all the typical annotations like `@Path`, `@GET` etc. then you'll see this failure when invoking the resource:
```
2020-10-25 23:39:01,178 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /hello failed, error id: fb45d7cc-5fa1-4aa5-ae5c-fa458d3b702d-1: org.jboss.resteasy.spi.UnhandledException: java.lang.RuntimeException: java.lang.InstantiationException
        at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:381)
        at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:131)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:37)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:94)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at java.base/java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Caused by: java.lang.RuntimeException: java.lang.InstantiationException
        at io.quarkus.arc.runtime.BeanContainerImpl$DefaultInstanceFactory.create(BeanContainerImpl.java:75)
        at io.quarkus.resteasy.common.runtime.QuarkusConstructorInjector.construct(QuarkusConstructorInjector.java:54)
        at org.jboss.resteasy.plugins.server.resourcefactory.POJOResourceFactory.createResource(POJOResourceFactory.java:71)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:386)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:68)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
        ... 17 more
Caused by: java.lang.InstantiationException
        at java.base/jdk.internal.reflect.InstantiationExceptionConstructorAccessorImpl.newInstance(InstantiationExceptionConstructorAccessorImpl.java:48)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.lang.Class.newInstance(Class.java:584)
        at io.quarkus.arc.runtime.BeanContainerImpl$DefaultInstanceFactory.create(BeanContainerImpl.java:67)
        ... 22 more
```
Such abstract base classes can be the result of https://github.com/OpenAPITools/openapi-generator

**Expected behavior**
Resource classes can use abstract base classes (that contain JAX-RS annotations) without problems.

**Actual behavior**
Invocation of such a resource fails.

**To Reproduce**

1. `git clone https://github.com/famod/quarkus-quickstarts`
2. `git switch rest-abstract`
3. `cd getting-started`
4. `mvn clean verify` (will fail with the mentioned error)
5. `mvn clean package -DskipTests`
6. `java -jar target/getting-started-1.0-SNAPSHOT-runner.jar`
7. `curl localhost:8080/hello` (will fail with the mentioned error)

**Configuration**
not relevant

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: `MINGW64_NT-10.0-18363 XXX 3.0.7-338.x86_64 2019-11-21 23:07 UTC x86_64 Msys`
 - Output of `java -version`: `OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.8+10, mixed mode)`
 - GraalVM version (if different from Java): n/a
 - Quarkus version or git rev: 1.9.0.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Apache Maven 3.6.3

**Additional context**
Interfaces do work but annotation discovery has some blind spots in that case:
- https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/jwt.20role.20handling
- #12736

### PR DESCRIPTION
**Pull request by patriot1burke**
Fix for issue #12927

### CODE REVIEW

**Comment by famod** [comment]
I am +1 for backporting this (1.9.1). WDYT?

**Review by kenfinnigan — APPROVED** [review]

**Review by geoand — APPROVED** [review]
:+1:

**Comment by geoand** [comment]
Yeah, I see no reason not to backport it

File: extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java [hunk 1]
```diff
@@ -279,6 +284,27 @@ public void build(
                 }
             }
         }
+        // look for all implementations of abstract classes annotated @Path
+        do {
+            Set<DotName> tmpAbstract = new HashSet<>();
+            for (final DotName iface : pathAbstract) {
+                final Collection<ClassInfo> implementors = index.getAllKnownSubclasses(iface);
```
Comment (gsmet) [inline extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java:291]: Given you get all the known subclasses here, I fail to see while you need the outer while loop?

This method also includes the non-direct subclasses.

File: extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java [hunk 2]
```diff
@@ -279,6 +284,27 @@ public void build(
                 }
             }
         }
+        // look for all implementations of abstract classes annotated @Path
+        do {
+            Set<DotName> tmpAbstract = new HashSet<>();
+            for (final DotName iface : pathAbstract) {
```
Comment (gsmet) [inline extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java:290]: Could we rename the variable? It's not an interface but an abstract class.

**Review by gsmet — DISMISSED** [review]
Added a suggestion and a question.

File: extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java (hunk 1 above)
Comment (patriot1burke) [reply extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java:291]: concrete class of subclass of subclass

File: extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java (hunk 2 above)
Comment (patriot1burke) [reply extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java:290]: No offense, but its often hard enough to get a PR approved with these kinds of style reviews.

**Comment by patriot1burke** [comment]
@geoand @kenfinnigan Please reapprove. Had minor fix proposed from Guillaume.

**Review by stuartwdouglas — APPROVED** [review]

**Review by kenfinnigan — APPROVED** [review]

**Pull request merged**